virtual_rainforest/main.py

```diff
@@ -108,20 +230,38 @@ def vr_run(
         "All models successfully configured, now attempting to initialise them."
     )
 
-    # This is just a step to pass flake8 checks (DELETE LATER)
-    print(models_cfd)
+    # Extract all the relevant timing details
+    start_time, end_time, update_interval = extract_timing_details(config)
+
+    # Identify models with shorter time steps than main loop and warn user about them
+    check_for_fast_models(models_cfd, update_interval)
 
     # TODO - Extract input data required to initialise the models
 
     # TODO - Initialise the set of configured models
 
-    # TODO - Save model state
+    # TODO - Spin up the models
+
+    # Start timing for all models
+    for model in models_cfd:
+        model.start_model_timing(start_time)
 
-    # TODO - Add timing loop
-    # TODO - Find models to update
-    # TODO - Solve models to steady state
     # TODO - Save model state
 
+    # Setup the timing loop
+    current_time = setup_timing_loop(start_time, end_time, update_interval)
+    while current_time < end_time:
+
+        current_time += update_interval
+
+        to_refresh, fixed = get_models_to_update(current_time, models_cfd)
+
+        # TODO - Solve models to steady state
+
+        models_cfd = to_refresh + fixed
```

Comment:
I'm not convinced by this - there is a lot of model passing around for what need to be some frequent checks. And I'm also not convinced by the separate   time 'grain' for the models and update_interval.

* I think update_interval should small (let's say 1 minute). There is a cost to repeating this loop at a fast frequency but all it should be doing really is keeping track of changing time.
* The whole `get_models_to_update` is _expensive_. I think we should track just a set of `next_update`times. Something like this:

```python
# Get the list of date times of the next update.
update_due = {mod.name: mod.next_update for mod in models_cfd}

# Loop until the end
while current_time < end_time:
    # Get the names of models that have expired due dates
    update_needed = [nm for nm, upd in update_due.items() if upd < current_time ]

    # If there are expired model, run their update() method and update their update due dates
    if update_needed:
        for mod_nm in update_needed:
            models_cfg[mod_nm].update()
            update_due[mod_nm] = models_cfg[mod_nm].next_update()

```

Reply:
There's a huge issue in here with models running at different rates and how to synchronize them. Ultimately that would be great to actually do but we need to kick this massively into touch for now and enforce synchronised model updates.

Reply:
Fair enough, but I suspect that cost will be trivial compared to the update() costs! I've just tried running the loop (including list comprehension due date check) using 1 hour intervals and a century took about 10 seconds, so minutes would probably be a bit much (guessing at 10 minutes).

Reply:
In [169]: print(end - start)
0:10:31.752243

Reply:
I like very much David's alternative. It should definitely be faster than copying models around. It can also be simplified slightly:

```python 
# Get the list of date times of the next update.
update_due = {mod.name: mod.next_update for mod in models_cfd}

# Loop until the end
while current_time < end_time:
    # Get the names of models that have expired due dates
    update_needed = [nm for nm, upd in update_due.items() if upd <= current_time ]

    # If there are expired model, run their update() method and update their update due dates
    for mod_nm in update_needed:
        models_cfg[mod_nm].update()
        update_due[mod_nm] = models_cfg[mod_nm].next_update()
```

Reply:
Oooh, yes. Entirely redundant `if` statement 😞